### ISSUE TITLE
Aggregate per-peer metriics

### ISSUE THREAD

**Issue by matklad** [near/nearcore#7596]
NB: SocketAddr is a label with an extremely high cardinality (~unbounded) potential. Given enough time the prometheus database will end up with a huge number of distinct timeseries which can lead to significant scaling issues with the metrics database.

_Originally posted by @nagisa in https://github.com/near/nearcore/issues/7530#issuecomment-1241911751_

---

I think the reasonable fix here would be to pre-aggregate the metric for all peers. Indeed, I split it per-peer simply because that  was easy to implement, but, after the recent rewrite, aggregation would be much easier.

**Comment by nagisa** [comment]
Well, so pardon my relative in experience with this area of the codebase, but it seems to me that nothing actually requires that the IP address for a node stays constant, or that the number of nodes is bounded, right?

> we explicitly delete metrics after peer gets disconnected

Do you mean we delete the series for a node from prometheus? Weird thing to do, but yeah, it would largely mitigate the issue(s) here, I think (provided there’s a “reasonable” limit on currently active/connected peers).

---

Side note: I see that the ports are constant today which helps this issue, but if they ever become auto-allocated, the situation would regress significantly. That’s just a symptom of a brittle system/infrastructure, so if there is an alternative option to a full `SocketAddr` (recording just an IP address part of it would be an improvement), I think that’d be ideal.

**Comment by pompon0** [comment]
> Well, so pardon my relative in experience with this area of the codebase, but it seems to me that nothing actually requires that the IP address for a node stays constant, nor the number of nodes is bounded, right?

Yes, there is no such requirement. I'm just saying that in practice the address from which prometheus collects metrics is as stable as the node P2P address. First address is a built-in label of the time series that prometheus stores, the second one is a custom one. I'm wondering whether they are treated differently.

> Do you mean we delete the series for a node from prometheus? 

What I mean is that we are removing locally a value from the GaugeVec, effectively terminating a time series since prometheus won't be able to collect any more data points for the given label.

> provided there’s a “reasonable” limit on currently active/connected peers

current limit is 40.

>  if there is an alternative option to a full SocketAddr (recording just an IP address part of it would be an improvement), I think that’d be ideal.

I don't think there is a strong need for recording any non-aggregated metric.

**Comment by nagisa** [comment]
> I'm just saying that in practice the address from which prometheus collects metrics is as stable as the node P2P address. First address is a built-in label of the time series that prometheus stores, the second one is a custom one. I'm wondering whether they are treated differently.

Well, this label is well bounded in practice, then this issue is a false alarm and probably can be closed ^^

---

> What I mean is that we are removing locally a value from the GaugeVec, effectively terminating a time series since prometheus won't be able to collect any more data points for the given label.

Ah, well that will only help with the collection performance and such, but if the prometheus database itself is storing many thousands of timeseries, its ability to query the datasets starts suffer. Each unique name/label/value combination is an unique time series, which is why the size of the label values’ set is a concern.

Having an aggressive retention policy helps. On the other hand, being able to query information about your systems long time into the past is super helpful when looking for correlations and such.

**Comment by pompon0** [comment]
> Having an aggressive retention policy helps. On the other hand, being able to query information about your systems long time into the past is super helpful when looking for correlations and such.

I don't know how flexible prometheus configuration is, but having a separate (aggressive) retention policy for metrics with high label cardinality in addition to a long retention policy for low cardinality metrics is a way to go. If prometheus supports that, we could have 2 metrics doing the same thing, one aggregating another, having separate retention policy.
